planner/core/push_down_predicate.go

```diff
@@ -0,0 +1,246 @@
+	}
+}
+
+/**
+ * @brief: This function is used transform the columns to a string of 0 and 1.
+ * @param: cols: the columns of a Expression
+ * @param: totalColumnCount: the total number of columns in the tablescan
+ * @return: the string of 0 and 1
+ */
+func transformColumnsToCode(cols []*expression.Column, totalColumnCount int) string {
+	code := make([]byte, totalColumnCount)
+	for _, col := range cols {
+		code[col.Index] = '1'
+	}
+	return string(code)
+}
+
+/**
+ *
+ * @brief: This function is used to group the conditions by the column they use
+ *         and sort the groups by the selectivity of the conditions in the group.
+ * @param: conds: the conditions to be grouped
+ * @return: the groups of conditions sorted by the selectivity of the conditions in the group
+ * @example: conds = [a > 1, b > 1, a > 2, c > 1, a > 3, b > 2]
+ *           return = [[a > 3, a > 2, a > 1], [b > 2, b > 1], [c > 1]]
+ * @note: when the selectivity of one group is larger than the threshold,
+ *  	  we will remove it from the returned result.
+ */
+func groupByColumnsSortBySelectivity(sctx sessionctx.Context, conds []expression.Expression, physicalTableScan *PhysicalTableScan) []expressionGroup {
+	// Create a map to store the groupMap of conditions keyed by the columns
+	groupMap := make(map[string][]expression.Expression)
+
+	// Iterate through the conditions and group them by columns
+	for _, cond := range conds {
+		columns := expression.ExtractColumns(cond)
+
+		var code string
+		if len(columns) == 0 {
+			code = "0"
+		} else {
+			code = transformColumnsToCode(columns, len(physicalTableScan.Columns))
+		}
+		groupMap[code] = append(groupMap[code], cond)
+	}
+
+	// Estimate the selectivity of each group and check if it is larger than the selectivityThreshold
+	var exprGroups []expressionGroup
+	for _, group := range groupMap {
+		selectivity, _, err := physicalTableScan.tblColHists.Selectivity(sctx, group, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		if selectivity <= selectivityThreshold {
+			exprGroups = append(exprGroups, expressionGroup{exprs: group, selectivity: selectivity})
+		}
+	}
+
+	// Sort exprGroups by selectivity in ascending order
+	sort.SliceStable(exprGroups, func(i, j int) bool {
+		return exprGroups[i].selectivity < exprGroups[j].selectivity || (exprGroups[i].selectivity == exprGroups[j].selectivity && len(exprGroups[i].exprs) < len(exprGroups[j].exprs))
+	})
+
+	return exprGroups
+}
+
+/**
+ * @brief: This function is used to check if the conditions contain heavy cost functions.
+ * @param: conds: some conditions of PhysicalSelection to be checked
+ * @note: heavy cost functions are functions that may cause a lot of memory allocation or disk IO.
+ * 		  For example, JSON functions.(To be supplemented)
+ */
+func withHeavyCostFunction(conds []expression.Expression) bool {
```

Comment:
Can we make the HeavyCost function configurable?

Reply:
no, HeavyCost function should be consistent rather than user-configurable